### DISCUSSION TITLE
v5 syntax to add preset users (in mnesia) using environment variables?

### DISCUSSION THREAD

**Discussion by atlantis** [emqx/emqx#8321] [Q&A] [answered]
Hi there! So in a docker container with the new version 5.0.0, I have authentication configured like so:

```authentication = [
  {
    backend = "built_in_database"
    mechanism = "password_based"
    password_hash_algorithm {name = "plain", salt_position = "suffix"}
    user_id_type = "username"
  }
]
```

And then I'm trying to add static users with environment variables... so I tried: 

`EMQX_AUTH__CLIENT__1__USERNAME: testuser`
`EMQX_AUTH__CLIENT__1__PASSWORD: testpass`

But the user isn't added. Is there another syntax I can use to accomplish this? Thanks!

**Comment by zmstone** [comment, marked as answer]
Hi @atlantis 
Client username and password etc is data, but not config. (Because there could be millions)
You can add users from dashboard or 
batch import from a csv file by calling a management API. (Find it in http://localhost:18083/api-docs/index.html)

edit: the api doc (swagger) isn't quite complete for this API, here is a quick curl example:
```
curl -v -u admin:public -X 'POST' \
    -H 'Content-Type: multipart/form-data' \
    -F 'filename=@/tmp/myusers.csv' \
    'http://localhost:18083/api/v5/authentication/password_based%3Abuilt_in_database/import_users'
```

**Reply by atlantis** [reply]
OK thanks! My use-case is deploying on Fargate, and so ideally if the service restarts it would load pre-build users so my app could connect immediately, but I'll find a workaround...

**Reply by zmstone** [reply]
this is an interesting requirement, maybe we can add a bootstrap command to import users from a pre-existing csv file.

**Reply by atlantis** [reply]
No worries - I'll just have to use two different fargate services for now (one container with persistence and the rest without persistence that connect to the persistent node to join the cluster).

**Reply by slaesh** [reply]
Any news on this @zmstone ? =)
I would need a mechanism too to bootstrap the container that there is atleast one valid user in the builtin-database :)

**Reply by slaesh** [reply]
just started a discussion to document this properly. :)

https://github.com/emqx/emqx/discussions/9142

**Reply by tvories** [reply]
I would also like some programmatic way to import users so I can git-ops my emqx configuration.

**Reply by simpsons310** [reply]
> this is an interesting requirement, maybe we can add a bootstrap command to import users from a pre-existing csv file.

Is this available now? For version `5.0.13`. And is there any plan to implement this soon? I'm using version `4.4.4` and going to upgrade to `5.x` version, but this feature is required for my application.

I think it would better if we can configure user credentials via environment variables like previous versions, so auto deployment can be easier in multi environment (dev, staging, prod, ...). Using `csv` file 

Thanks.
